docs: Rework the Build your first app guide for 3.5 (#557)
serverpod/serverpod_docs: docs/05-build-your-first-app/02-models-and-data.md

---
title: Create data models
sidebar_class_name: sidebar-icon-get-started-step-2
slug: /get-started/models-and-data
description: Define a Serverpod data model so your endpoint returns typed, structured data, with serialization between server and client generated for you.
---

<!-- markdownlint-disable MD025 -->

# Create data models

> On the [previous page](./01-creating-endpoints.md) your endpoint returned a plain string. Here you'll define a `Recipe` model so the server returns structured, typed data instead. You define the model once in YAML, and Serverpod generates the Dart class plus all the serialization that moves it between server and client.

Reviewer: ```suggestion
On the [previous page](./01-creating-endpoints.md) your endpoint returned a plain string. Here you'll define a `Recipe` model so the server returns structured, typed data instead. You define the model once in YAML, and Serverpod generates the Dart class plus all the serialization needed between server and client.
```

:::

## Return the model from your endpoint

Now update `recipe_endpoint.dart` to return a `Recipe` instead of a `String`.

Import the generated models at the top of the file:

```dart
import '../generated/protocol.dart';
```

Change the method's return type from `Future<String>` to `Future<Recipe>`:

```dart
  Future<Recipe> generateRecipe(Session session, String ingredients) async {
```

Then replace `return responseText;` with a constructed `Recipe`:

```dart
    return Recipe(
      author: 'Gemini',
      text: responseText,
      date: DateTime.now(),
      ingredients: ingredients,
    );
```

Save the file. The client regenerates so `generateRecipe` now returns a `Recipe`.

## Show it in your app

In `recipe_screen.dart`, the call now returns a `Recipe` object, so assigning `result` to a string no longer compiles. `_resultMessage` itself stays a `String?`; you just format the `Recipe`'s fields into it. In `_callGenerateRecipe`, update the success `setState`:

```dart
      setState(() {
        _resultMessage = '${result.author} on ${result.date}:\n${result.text}';
        _loading = false;
      });
```

> Save. Because the model and the endpoint's return type changed, the generated client changed too, so hot reload isn't enough. Press `R` in the `serverpod start` terminal to hot restart.

Reviewer: This Press R action is also easily missed.

Enter some ingredients and tap **Generate Recipe**. The result now shows the author and date alongside the recipe:

> ![Flutter Recipe App](/img/getting-started/flutter-web-ingredients.png)

Reviewer: We need a new screenshot. The button says "Send to server", while the code and guide set it to "Generate Recipe".

> ![Flutter Recipe App](/img/getting-started/flutter-web-ingredients.png)

Reviewer: The screenshot does not show the author, not even the result, as stated in the line above. Also, consider zooming in a bit in the browser for the screenshot to make sure author and date is readable.
